jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt

```diff
@@ -199,28 +208,56 @@ class DefaultAwsResourceCache(private val project: Project, private val clock: C
         forceFetch: Boolean
     ): CompletionStage<T> = when (resource) {
         is Resource.View<*, T> -> getResource(resource.underlying, region, credentialProvider, useStale, forceFetch).thenApply { resource.doMap(it as Any) }
-        is Resource.Cached<T> -> Context(resource, region, credentialProvider, useStale, forceFetch).also { getCachedResource(it, 0) }.future
+        is Resource.Cached<T> -> Context(resource, region, credentialProvider, useStale, forceFetch).also { getCachedResource(it) }.future
     }
 
-    private fun <T> getCachedResource(context: Context<T>, failedAttempts: Int) {
+    private fun <T> getCachedResource(context: Context<T>) {
         ApplicationManager.getApplication().executeOnPooledThread {
             try {
                 @Suppress("UNCHECKED_CAST")
                 val result = cache.compute(context.cacheKey) { _, value ->
                     fetchIfNeeded(context, value as Entry<T>?)
                 } as Entry<T>
                 context.future.complete(result.value)
-            } catch (e: AssertionError) { // HACK: until we figure out why Guava Cache is throwing an exception : https://github.com/google/guava/issues/3464
-                when {
-                    failedAttempts > 0 -> context.future.completeExceptionally(e)
-                    else -> getCachedResource(context, failedAttempts + 1)
-                }
             } catch (e: Throwable) {
                 context.future.completeExceptionally(e)
             }
         }
     }
 
+    private fun runCacheMaintenance() {
+        try {
+            val entries = cache.entries.toList()
+            val totalWeight = entries.sumBy { it.value.weight }
+            val exceeded = totalWeight - maximumCacheEntries
+            if (exceeded <= 0) return
+            val sortedValues = entries.sortedBy { it.value.expiry }
+            prune(sortedValues, 0, exceeded)
+        } finally {
+            scheduleCacheMaintenance()
+        }
+    }
+
+    private fun scheduleCacheMaintenance() {
+        if (!alarm.isDisposed) {
+            alarm.addRequest(this::runCacheMaintenance, maintenanceInterval.toMillis())
+        }
+    }
+
+    private tailrec fun prune(entries: List<Map.Entry<CacheKey, Entry<*>>>, removedSoFar: Int, targetRemoval: Int) {
+        if (removedSoFar >= targetRemoval) return
+        val head = entries.firstOrNull() ?: return
+        val nowRemoved = if (removeIfStillSame(head)) {
+            removedSoFar + head.value.weight
+        } else {
+            removedSoFar
+        }
+        prune(entries.tail(), nowRemoved, targetRemoval)
+    }
+
+    private fun removeIfStillSame(head: Map.Entry<CacheKey, Entry<*>>): Boolean =
+        cache.computeIfPresent(head.key) { _, currentEntry -> if (currentEntry == head.value) null else currentEntry } == null
```

Comment:
This should be an explicit identity check instead of ==

Reply:
yup - updated